### PR TITLE
fix(sharding): Improve sharding async migration

### PR DESCRIPTION
**Pull request by macobo**
## Changes

This PR is a follow-up to https://github.com/PostHog/posthog/pull/8783, solving left-over issues from that PR.

Read https://github.com/PostHog/posthog/issues/8652 for full context on the issue

Note I did _not_ update zookeeper automatically - I don't think it'll become an issue in practice, but if it does we'll need to follow up anyways. Automatically clearing data from zookeeper runs into the risk of dropping real user data.

## How did you test this code?

Unit tests, running the migration manually succeeds if env is set.

### CODE REVIEW

File: ee/clickhouse/sql/test/__snapshots__/test_schema.ambr [hunk 1]
```diff
@@ -150,6 +150,9 @@
       created_at DateTime64(6, 'UTC')
       
       
+  , _timestamp DateTime
```
Comment (yakkomajuri) [inline ee/clickhouse/sql/test/__snapshots__/test_schema.ambr:153]: unrelated to the scope but we're still defaulting to `_offset` without `_partition`. I guess it's somewhat helpful but not entirely

**Review by yakkomajuri — APPROVED** [review]
nice, I like this.

thanks for taking the time. this will serve as a good example of safe rollbacks.

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 2]
```diff
@@ -87,12 +87,9 @@ class Migration(AsyncMigrationDefinition):
     depends_on = "0003_fill_person_distinct_id2"
 
     def is_required(self):
-        return "Distributed" not in self.get_current_engine("events")
+        return "Distributed" not in cast(str, self.get_current_engine("events"))
 
     def precheck(self):
```
Comment (tiina303) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:92]: Add to the pre-check a check about backup table existing similar to https://github.com/PostHog/posthog/pull/8556

File: posthog/async_migrations/migrations/0004_replicated_schema.py [hunk 3]
```diff
@@ -217,6 +225,25 @@ def move_partitions(self, from_table: str, to_table: str):
             sync_execute(f"ALTER TABLE {to_table} ATTACH PARTITION {partition} FROM {from_table}")
             sync_execute(f"ALTER TABLE {from_table} DROP PARTITION {partition}")
 
+    def rename_tables(
```
Comment (tiina303) [inline posthog/async_migrations/migrations/0004_replicated_schema.py:228]: we might want to have this in some sort of utility functions file for async migrations, but no need to do that now.

File: posthog/async_migrations/migrations/0004_replicated_schema.py (hunk 2 above)
Comment (tiina303) [reply posthog/async_migrations/migrations/0004_replicated_schema.py:92]: actually ignore this you are renaming it back to the tmp name which is different from what we did in events to keep the failed attempt around

**Review by tiina303 — APPROVED** [review]

**Pull request merged**